### PR TITLE
fix(payments): write a foreign_from implementation for payment_method_data and add missing payment methods in helpers.rs

### PR DESCRIPTION
**Pull request by kritikmodi**
## Type of Change

- [x] Bugfix

## Description

This is a follow-up change for https://github.com/juspay/hyperswitch/pull/1236. This adds a `foreign_from` implementation for `payment_method_data` for the payment methods correspondence check.

## Motivation and Context

This PR closes https://github.com/juspay/hyperswitch/issues/1800.

## How did you test it?

Manually using Postman.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code

### CODE REVIEW

**Review by Sangamesh26 — APPROVED** [review]
LGTM :)

**Review by prajjwalkumar17 — DISMISSED** [review]
LGTM

File: crates/router/src/core/payments/helpers.rs [hunk 1]
```diff
@@ -1434,6 +1434,18 @@ pub(crate) fn validate_payment_method_fields_present(
...
+                    api_enums::PaymentMethod::BankTransfer,
+                    api::PaymentMethodData::BankTransfer(..)
+                ) | (
+                    api_enums::PaymentMethod::Reward,
+                    api::PaymentMethodData::Reward(..)
+                ) | (
+                    api_enums::PaymentMethod::Upi,
+                    api::PaymentMethodData::Upi(..)
+                ) | (
+                    api_enums::PaymentMethod::Voucher,
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments/helpers.rs:1447]: How are we going to ensure that a new payment method would be added here too?

Comment (kritikmodi) [reply crates/router/src/core/payments/helpers.rs:1447]: Right. I'll add a `foreign_from` implementation to ensure this.

File: crates/router/src/core/payments/helpers.rs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 use std::borrow::Cow;
 
+// use api_models::enums::PaymentMethod;
```
Comment (Sangamesh26) [inline crates/router/src/core/payments/helpers.rs:3]: Remove commented line

**Review by Sangamesh26 — CHANGES_REQUESTED** [review]

File: crates/router/src/types/transformers.rs [hunk 3]
```diff
@@ -230,6 +230,30 @@ impl ForeignFrom<api_enums::PaymentMethodType> for api_enums::PaymentMethod {
...
+            api_models::payments::PaymentMethodData::BankTransfer(..) => Ok(Self::BankTransfer),
+            api_models::payments::PaymentMethodData::Crypto(..) => Ok(Self::Crypto),
+            api_models::payments::PaymentMethodData::Reward(..) => Ok(Self::Reward),
+            api_models::payments::PaymentMethodData::Upi(..) => Ok(Self::Upi),
+            api_models::payments::PaymentMethodData::Voucher(..) => Ok(Self::Voucher),
+            api_models::payments::PaymentMethodData::GiftCard(..) => Ok(Self::GiftCard),
+            _ => Err(errors::ApiErrorResponse::InvalidRequestData {
+                message: ("Mandate payments cannot have payment_method_data field".to_string()),
+            }),
+        }
```
Comment (SamraatBansal) [inline crates/router/src/types/transformers.rs:253]: How this `_` case will come when `payment_method_data` is an enum.
Also are we checking things like `off_session:true` before classifying it as mandate payment

File: crates/router/src/core/payments/helpers.rs (hunk 2 above)
Comment (kritikmodi) [reply crates/router/src/core/payments/helpers.rs:3]: Okay

File: crates/router/src/types/transformers.rs (hunk 3 above)
Comment (kritikmodi) [reply crates/router/src/types/transformers.rs:253]: Resolved

File: crates/router/src/core/payments/helpers.rs [hunk 4]
```diff
@@ -1416,40 +1416,31 @@ pub(crate) fn validate_payment_method_fields_present(
...
+                            }
+                        }
+                        Err(err) => Err(err),
+                    },
+                    None => Ok(()),
+                }
+            }
+            None => Ok(()),
         },
     )?;
```
Comment (jarnura) [inline crates/router/src/core/payments/helpers.rs:1445]:
```suggestion
    let check_payment_method_and_payemnt_method_data =
        |req_payment_method_data, req_payment_method: api_enums::PaymentMethod| {
            api_enums::PaymentMethod::foreign_try_from(req_payment_method_data).and_then(|payment_method|
                if req_payment_method != payment_method {
                    Err(errors::ApiErrorResponse::InvalidRequestData {
                        message: ("payment_method_data doesn't correspond to the specified payment_method"
                            .to_string()),
                    })
                } else {
                    Ok(())
                })
        };

    utils::when(
        req.payment_method.is_some() && req.payment_method_data.is_some(),
        || {
            req.payment_method_data
                .clone()
                .map_or(Ok(()), |req_payment_method_data| {
                    req.payment_method.map_or(Ok(()), |req_payment_method| {
                        check_payment_method_and_payemnt_method_data(
                            req_payment_method_data,
                            req_payment_method,
                        )
                    })
                })
        },
    )?;
```

Idiomatic way

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments/helpers.rs (hunk 4 above)
Comment (kritikmodi) [reply crates/router/src/core/payments/helpers.rs:1445]: Okay

File: crates/common_enums/src/enums.rs [hunk 5]
```diff
@@ -954,6 +954,7 @@ pub enum PaymentMethod {
     Reward,
     Upi,
     Voucher,
+    GiftCard,
```
Comment (Narayanbhat166) [inline crates/common_enums/src/enums.rs:957]: Why is this added in this PR?

**Review by lsampras — APPROVED** [review]

File: crates/router/src/types/transformers.rs [hunk 6]
```diff
@@ -230,6 +230,176 @@ impl ForeignFrom<api_enums::PaymentMethodType> for api_enums::PaymentMethod {
     }
 }
 
+impl ForeignTryFrom<(Self, api_enums::PaymentMethodType)> for api_enums::PaymentMethod {
```
Comment (vspecky) [inline crates/router/src/types/transformers.rs:233]: Why do you need a `ForeignTryFrom`? You use `ForeignTryFrom` for *conversions* when you want to convert from a type `T` to a type `U`. In this case you already have both `T` and `U`. You're not converting here. You're validating. A simple function for that should suffice

**Review by vspecky — CHANGES_REQUESTED** [review]

**Review by vspecky — APPROVED** [review]

File: crates/router/src/types/transformers.rs (hunk 6 above)
Comment (vspecky) [reply crates/router/src/types/transformers.rs:233]: We'll pick this up in another PR

**Review by vspecky — DISMISSED** [review]

**Review by jarnura — APPROVED** [review]

**Review by Sangamesh26 — APPROVED** [review]
